### ISSUE TITLE
Vector empty base class optimization for allocator

### ISSUE THREAD

**Issue by remyabel** [electronicarts/EASTL#99]
Noticed this comment in [vector.h#L156](https://github.com/electronicarts/EASTL/blob/master/include/EASTL/vector.h#L156)
```
allocator_type mAllocator; // To do: Use base class optimization to make this go away.
```

But you already have a [`compressed_pair`](https://github.com/electronicarts/EASTL/blob/master/include/EASTL/bonus/compressed_pair.h) class.

So it would be trivial to do this and get the EBCO for free:
```
compressed_pair<T*, allocator_type> allocator_pair;
```

Though this requires refactoring in many places. i.e [vector.h#L437](https://github.com/electronicarts/EASTL/blob/master/include/EASTL/vector.h#L437)

Replace:

```
mpCapacity(NULL),
mAllocator(EASTL_VECTOR_DEFAULT_NAME)
```
with:
```
allocator_pair(nullptr, EASTL_VECTOR_DEFAULT_NAME)
```

**Comment by rparolin** [comment]
I'm working on this internally.  Stay tuned, hoping for those changes to be included in the next release.

**Comment by rparolin** [comment]
Quick update.  I actually did the EBCO for eastl::string, but I will try to squeeze in a similar update for vector in this or maybe the next release.  Doing the EBCO for eastl::string actually had significant knock-ons so I think I'll try to space out the edits to these heavily used containers.  :)

**Comment by remyabel** [comment]
The alternative to using `compressed_pair` is to factor out the implementation details into its own EBCO-enabled class like [libstdc++ does](https://github.com/remyabel/cpp-notes#libstdcs-approach). I'd be interested to see which approach is less "painful".

**Comment by rparolin** [comment]
libstdc++ solution seems to avoid building a compressed_pair (according to the article) but I'd argue that using a compressed_pair the intent is more clear, especially since both solutions still require code changes in vector itself.

**Comment by rparolin** [comment]
Resolved with release 3.05.06.
